### ISSUE TITLE
Admin login broken after update to magento 1.9.4.3

### ISSUE THREAD

**Issue by tobihille** [ikonoshirt/pbkdf2#31]
Hi,

it is not possible to log into admin, as it seems the hash comparison is failing. After password reset the 'password' - column in admin_user has a rather short hash (but longer than md5).

The shop was on 1.9.4.2 with pbkdf installed, then we upgraded to 1.9.4.3 wich lead to the described problems.

We managed to get it fixed via an ugly core hack (enough for friday evening), bit we're still looking for a nice solution.

- Tobi

**Comment by Schrank** [comment]
@tobihille Can you give some more infos? What hack? What is the problem? :-)

**Comment by Schrank** [comment]
The problem is, that Magento added bcrypt to admin logins with 1.9.4.3 and updated the password field to varchar(255) which can hold a bcrypt hash, but no pbkdf2.

Unfortunately this doesn't apply to customer passwords.

**Comment by mobweb** [comment]
@Schrank Hi Fabian!

I don't understand exactly, is the issue fixed now with the latest commits?

I have two instances, one where I've already updated to 1.9.4.4 and now can not log in to the Admin Panel anymore, even after installing the latest version of your extension.

On another instance, I have not yet updated to 1.9.4.4. Can I assume no issues if I first install the latest version of your extension and then update Magento?

Thanks!

**Comment by Schrank** [comment]
Admin user is fixed. API seems to be still broken. Worked locally but not live :-/

**Comment by Schrank** [comment]
@mobweb sorry didn'T read correctly. The issue is fixed, but the Magento update changes the DB table to VARCHAR(255) which doesn't happen now. If you already updated the Magento and not PBKDF2, you need to change the database back and the password hashes are broken - aka reset passwords for alle admin users :-/

**Comment by mobweb** [comment]
Ah, I see, thanks for the clarification. It's no big deal as we only broke the passwords locally. :)

If I do not have a backup though, could I not simply change the definition of the password field back to "Text" without a Length, and then reset my passwords to be able to log back in?

Actually, I did just that, and its not letting me log in. And when I create a new admin user or change my password via Magerun, its still writing a rather short hash into the password field, something like this:

$2y$10$IuC0Rs0y8SU7QTp4ysIy/e3.WTLw4lJ2u3syTcdorpIcrN0N7/KMG

But as I understand it after changing the field back to Text without a Length and creating a new admin user, the password hash should be quite long again, much longer than the one I posted above, right?

EDIT: Also tested using Magento's reset password functionality now. Hash written into the Password field is very short, much like the one I posted above. Am I missing something here? Thanks. :)

EDIT 2: Another test I just did: Disabled your extension and then reset the PW via the email/form. The hash written still looks the same, but now its letting me log in, while your extension is still disabled. So to me it looks like the hashes being generated are using Magento's core hashing functionality, even when your extension is enabled. But this might be an edge case for my installation as we did some weird stuff before, so not sure if you would be able to reproduce this bug.

**Comment by Schrank** [comment]
This is a bcrypt hash, it is a "good hash" imho (I'm no security expert!) but it is FAR better than the md5 bullshit Magento used before. And if used correctly (that's the part I don't know) it is a proper replacement for pbkdf2.

So without the pbkdf2 update admins use crypt now and customers imho too. (didn't test it), so maybe you can just go with the current version.

The worst case is, login is broken for customers too, which I expect.

**Comment by mobweb** [comment]
Ah I see. So I'll try to get it running properly with the bcrypt hashes. I fixed the admin logins by simply disabling your extension and recreating the passwords for the admin users. That is fine for me.

Now I will see what I can find out about the customer accounts and let you know how it works out for me.

Thanks!

**Comment by Schrank** [comment]
You can't deactivate the extension without reseting all customer passwords :-/

What you can do is implement the same as this extension: A migration from pbkdf2 to bcrypt on login. You know the password on login :-) But there is no way to migrate the passwords instantly. Once you migrated "most" of the passwords you can deactivate the extension and tell all other customers to reset them (if they didn'T show up for the last n months)

**Comment by mobweb** [comment]
Ah, I see. Well the good news is that the customer passwords don't seem to be affected by this. Was just able to log into an old account while your extension was enabled on 1.9.4.4.

But for some reason, admin logins don't seem to work for me when the extension is enabled. I "fixed" the password field by changing it back to Text in the DB and then reset my password using the email/form. The password was then written into the DB as a bcrypt hash.

You are saying that I should now be able to login using this new password (which is stored as bcrypt in the DB), right?

For some reason, its always giving me a wrong password error message, until I disable the extension, then it is working. But that is also not an option, as the customers won't be able to log in then. ;)

**Comment by Schrank** [comment]
TL;DR for everyone else, you have a mix of the following problems:
- admin user password field in db is too short
- you have mixed passwords for admin/customer/api of the formats:
  - pbkdf2 (looks super long)
  - md5 (short)
  - bcrypt (starts with $y3$ or something alike)
